web/src/typeahead_helper.ts

```diff
@@ -270,17 +270,21 @@ export function compare_people_for_relevance(
         }
     }
 
-    // give preference to direct message partners if both (are)/(are not) subscribers
-    const a_is_partner = pm_conversations.is_partner(person_a.user_id);
-    const b_is_partner = pm_conversations.is_partner(person_b.user_id);
+    const preference = tertiary_compare(person_a, person_b);
```

Comment:
So notably, this change adjusts the order for DM conversations too! Previously, we first checked `pm_conversations.is_partner`, which just looks at whether we have a DM conversation with this person, and then `people.compare_by_pms`, which sorts by the number of DMs we've had with this person.

I think `people.compare_by_pms` will almost always be a more refined version of `pm_conversations.is_partner`, so maybe that's fine. But we should certainly call that out as what we're doing!

Hmm, I think they are doing different things, though -- `pm_conversations` looks at conversations, so an entry will be there for a group DM conversation, but not for either individual user, while `people.compare_by_dms` does the opposite -- it just counts how many DMs in the local cache include the user. I think probably the structure we actually want is:
- We rename `tertiary_compare` to `compare_by_current_conversation` and make that parameter optional; the `!current_stream` code path in the caller would just not pass it.
- This function uses `people.compare_by_pms` rather than `pm_conversations.is_partner` as a first tiebreak, since that seems somewhat correct for a "How relevant is this person to you" check than just if you've had a 1:1 DM conversation with them in our data set.
- We then just unconditionally check `pm_conversations.is_partner` after that -- this will almost never do anything, but should be more correct in the rare case that you have no recent DM message history with the target user, but do have some older message history that's outside the "recent messages only" data set powering `people.compare_by_pms`.

Those last two things might be improved upon by just tweaking the `pm_conversations` module to maintain a data set mapping `user => max_message_id` for conversations containing the user and just doing that one check, but that feels like it'd prove expensive to do with the current data structures. Might be worth leaving a comment noting that possibility.

Reply:
Thank you for noticing and bringing up the subtle difference in the two functions!

`typeahead_helper.compare_by_pms` does check for how many DMs include the users via `people.get_recipient_count`, but as a tie breaker, it first gives preference to human users, and then to the lexicographically smaller user. This means it practically will never return 0; the alphabetical tie breaker in this function will abort `compare_people_for_relevance` before it gets to the final `pm_conversations.is_partner` tie breaker.

So instead of having `pm_conversations.is_partner` as the final tie breaker in `compare_people_for_relevance`, I think it might be better to have it as a part of `typeahead_helper.compare_by_pms`, after the bot check and before the lexicographic tie breaker. Though this would change search suggestions as well, not only the recipient typeahead suggestions, I think it'd be a minor and sensible change. What do you think?

Reply:
@timabbott I have reworked the entire PR... Please take a look!